File: go/hashmap/242_is_anagram.go
package hashmap

// isAnagram 字母异位词
func isAnagram(s string, t string) bool {

	if len(s) != len(t) {
		return false
	}

	// 建立两个 hashMap 进行比较
	hmaps := make(map[rune]int)
	hmapt := make(map[rune]int)

	for _, v := range s {
		hmaps[v] += 1
	}

	for _, v := range t {
		hmapt[v] += 1
	}

	if len(hmaps) == len(hmapt) {
		for k := range hmaps {
			if hmaps[k] != hmapt[k] {
				return false
			}
		}
		return true
	}
	return false
}

File: rust/src/hashmap/_242_is_anagram.rs
pub fn is_anagram(s: String, t: String) -> bool {
    // HashMap
    use std::collections::HashMap;

    let s: Vec<char> = s.chars().collect();
    let t: Vec<char> = t.chars().collect();

    let mut hmap_s = HashMap::new();
    let mut hamp_t = HashMap::new();

    for ch in s {
        *hmap_s.entry(ch).or_insert(0) += 1;
    }

    // println!("s{:?}", s);

    for ch in t {
        *hamp_t.entry(ch).or_insert(0) += 1;
    }

    hamp_t == hmap_s
}

pub fn is_anagram2(s: String, t: String) -> bool {
    // HashMap  用一个HashMap
    use std::collections::HashMap;

    let s: Vec<char> = s.chars().collect();
    let t: Vec<char> = t.chars().collect();

    let mut hmap = HashMap::new();

    for ch in s {
        *hmap.entry(ch).or_insert(0) += 1;
    }

    // println!("s{:?}", s);

    for ch in t {
        *hmap.entry(ch).or_insert(0) -= 1;
    }

    !hmap.iter().any(|(_, &v)| v != 0)
    // hmap.iter().all(|(_, x)| *x == 0)
}
